src/coreclr/jit/morph.cpp

```diff
@@ -15503,3 +15503,212 @@ PhaseStatus Compiler::fgMorphArrayOps()
+// will be used as a mask throughout the code. This is the case that should be optimised for.
+//
+// Operation:
+// Look for a LCL which where the input is converted from a MASK. Remove the conversion and
+// updated the type to MASK. Find all uses of the LCL. For each use, update the type to MASK.
+// If it is converted to a MASK, then remove the node. Otherwise, add a conversion from vector
+// to mask.
+//
+// Arguments:
+//      stmt - Statement to check
+//
+// Returns:
+//    True if changes were made
+//
+bool Compiler::ConvertLCLMasks(Statement* stmt)
+{
+    // Look for:
+    //      STORELCL(TYP_SIMD, ConvertMaskToVector(mask))
+
+    GenTree* tree = stmt->GetRootNode();
+
+    if ((tree->OperGet() != GT_STORE_LCL_VAR) || (!varTypeIsSIMD(tree->gtType)) ||
+        (!tree->AsLclVar()->Data()->OperIsConvertMaskToVector()))
+    {
+        return false;
+    }
+
+    JITDUMP("Found Local mask store with conversion\n");
+    gtDispTree(tree);
+
+    GenTreeHWIntrinsic* convertOp = tree->AsLclVar()->Data()->AsHWIntrinsic();
+    unsigned const      lclNum    = tree->AsLclVarCommon()->GetLclNum();
+    GenTree*            maskOp    = convertOp->Op(1);
+
+    // Update the type of the STORELCL - including the lclvar.
+    tree->gtType      = maskOp->gtType;
+    LclVarDsc* varDsc = lvaGetDesc(lclNum);
+    varDsc->lvType    = maskOp->gtType;
+
+    // Remove the convert from the tree.
+    convertOp->gtBashToNOP();
+    tree->AsOp()->gtOp1 = maskOp;
+    fgSequenceLocals(stmt);
+
+    JITDUMP("\nRemoved conversion\n");
+    gtDispTree(tree);
+
+    // Find all uses of the LCL. These could be anywhere in the current method.
+    for (BasicBlock* block : Blocks())
+    {
+        for (Statement* const stmt : block->Statements())
+        {
+            for (GenTreeLclVarCommon* lcl : stmt->LocalsTreeList())
+            {
+                if (lcl->OperIs(GT_LCL_VAR) && (lcl->GetLclNum() == lclNum) && (lcl->gtType != TYP_MASK))
+                {
+                    JITDUMP("\nFound a use\n");
+                    gtDispTree(lcl);
+
+                    // Find the parent. If it is a ConvertVectorToMask then remove it.
+                    RemoveLCLUseConvertVisitor ev(this, lcl->GetLclNum(), stmt);
+                    GenTree*                   root = stmt->GetRootNode();
+                    ev.WalkTree(&root, nullptr);
+
+                    // TODO: If a ConvertVectorToMask was not found then insert a ConvertMaskToVector
+                    if (!ev.removedConversion)
+                    {
+                        assert(false);
+                    }
+                }
+            }
+        }
+    }
+
+    return true;
+}
+
+//------------------------------------------------------------------------
+// optLCLMasks: Allow locls to be of MASK type
+//
+// Returns:
+//    Suitable phase status
+//
+PhaseStatus Compiler::optLCLMasks()
+{
+    // TODO: Use a different config option
+#ifdef DEBUG
+    if (optConfigDisableCSE())
```

Comment:
This phase also applies to other architectures with masking support, it should not be dependent on SVE

Reply:
> This phase also applies to other architectures with masking support

I've not really looked at AVX512 dumps. Am I right in thinking this same issue exists on AVX512 but it's not as serious because there are fewer uses of masks?

Reply:
I definitely agree with @kunalspathak that we should leave some breadcrumb around to skip this phase. This phase is not going to do anything in vast majority of cases.

Reply:
> I definitely agree with @kunalspathak that we should leave some breadcrumb around to skip this phase. This phase is not going to do anything in vast majority of cases.

This could be as simple as a bool which is reset at the start of each method, and set in `gtNewSimdCvtMaskToVectorNode()`.

Reply:
> This phase is not going to do anything in vast majority of cases.

Right, don't disagree here. I also don't mind if the initial PR is only enabled for Arm64 since it's being done by our friends at Arm here.

My comment was rather that this phase is ultimately general purpose to any IR using `TYP_MASK` on any platform and so we shouldn't author it to be restricted unnecessarily. We should rather ensure it works on the general concepts like checking for and handling `TYP_MASK` instead.